docs/website/docs/extensions/index.md

```diff
@@ -0,0 +1,442 @@
+* :x: External runtime interactions (file/network/user IO). Use custom modules.
+
+### Implementation
+
+To make use of this approach one just needs to follow the standard
+[MLIR dialect conversion](https://mlir.llvm.org/docs/DialectConversion/)
+behavior: add a dialect with ops, add a conversion pass, and run that pass
+before providing the resulting IR to the IREE compiler. See
+[Creating a Dialect](https://mlir.llvm.org/docs/Tutorials/CreatingADialect/).
+
+Think of this like authoring C++ sources with templates that you compile into
+your application: Clang (and LLVM beyond) don't know about your library details
+and instead just process it as it would any other code. You can take the same
+source and pass it to GCC and it'll be robust to underlying changes in the
+system.
+
+## 2. Extend host code with custom modules
+
+* TL;DR: import MLIR functions in the compiler and custom modules at runtime.
+
+```mlir
+// Main user module compiled by IREE:
+module @model {
+  func @predict() {
+    %4 = call @my_custom_module::@some_func(%3) : (tensor<?xf32>) -> i32
+    ...
+  }
+}
+// External module that will be available at runtime:
+module @my_custom_module {
+  func @some_func(%input: tensor<?xf32>) -> i32  // note empty for extern
+}
+```
+
+IREE provides dynamic linking at runtime via its VM interfaces. For code that
+runs on the host and requires syscalls or calling out to existing libraries -
+such as file IO, text processing, and JPEG decoding - this is an easy way to
+interop without paying attention to the more complex details of device code.
+An IREE module compiled using custom modules is portable and dynamically
+deployable so long as the custom module is registered at runtime.
+
+This approach conceptually matches what normal native binaries do in an OS:
+imports are declared and at runtime they are resolved based on the available
+exports of modules in the system. Just as with normal systems engineering
+design of the API between modules is up to the user and depending on rigor can
+have several pitfalls but these problems and their solutions are not IREE
+specific and anyone who has designed a shared library interface can apply the
+same rules here in IREE around versioning, performance, etc. One does not add 2
+integers via a syscall and the same holds here: custom modules and the functions
+within should perform a large amount of work to hide overheads involved in the
+cross-module calls and users must be aware that the compiler cannot optimize
+across the call boundaries.
+
+### Pros
+
+* No IREE compiler code changes required.
+* Produced artifacts are portable across IREE deployment configurations.
+* Full system access is allowed - the VM just calls external functions.
+* Runtime modules can be implemented (via shims) in other languages/runtimes.
+
+### Cons
+
+* _(Today)_ no asynchronous scheduling or pipelining.
+  * Future improvements will enable pipelining of host calls.
```

Comment:
nit: mkdocs nested lists should have four spaces of indentation between levels
```suggestion
* _(Today)_ no asynchronous scheduling or pipelining.
    * Future improvements will enable pipelining of host calls.
```

(throughout this page)